### PR TITLE
[11.x] Enhance database migrations

### LINKED ISSUE #25475: [5.7] Sqlite & dropForeign
- Laravel Version: 5.7.1
- PHP Version: 7.2
- Database Driver & Version: MySQL 8, Sqlite

### Description:

When using Laravel, using in-memory Sqlite for tests while using MySQL/Postgres for dev/prod requests is quite common.

Laravel 5.7 contains an undocumented change which throws an exception when trying to drop foreign keys in Sqlite: https://github.com/laravel/framework/pull/24052 

When a testsuite runs migrations, it produces the message:
`SQLite doesn't support dropping foreign keys (you would need to re-create the table).`

While this is technically correct, this strict behavior also removes the possibility to silently ignore foreign keys in Sqlite and use the same migrations with multiple drivers.

Could the checks in https://github.com/laravel/framework/pull/24052 be reverted, or be made configurable? 

If not, I think it should at least be added to the 5.7 upgrade docs to set expectations. Rewriting hundreds of migrations is quite impactful.

### LINKED ISSUE #51318: Laravel 11 SQLite Migration Failure: after drop column: unknown column "XXXXXX" in foreign key definition
### Laravel Version

11.6.0

### PHP Version

8.3.6

### Database Driver & Version

SQLite 3.45.1

### Description

In Laravel 11, running tests in memory using SQLite I receive this error when trying to drop a column that have a foreign key on it. That wasn't a problem before when using `doctrine/dbal` (#48864).

### Steps To Reproduce

1. Create a new Laravel project.
2. Add a new migration that drop the `user_id` column on the `sessions` table like so:
```
Schema::table('sessions', function (Blueprint $table) {
    $table->dropColumn('user_id');
});
```
3. Setup PHPUnit for testing with the connection set to SQLite and the database set to :memory:.
4. Add `use RefreshDatabase;` in the default `TestCase` class
5. Run the test `test_the_application_returns_a_successful_response`.
6. You should receive the following error:
```
There was 1 error:

1) Tests\Feature\ExampleTest::test_the_application_returns_a_successful_response
Illuminate\Database\QueryException: SQLSTATE[HY000]: General error: 1 error in index sessions_user_id_index after drop column: no such column: user_id (Connection: sqlite, SQL: alter table "sessions" drop column "user_id")
```

### PR DESCRIPTION
**Pull request by hafezdivandari**
This PR does a lot of improvements and unlocks many possibilities on database migrations, also makes it easier to maintain!

### Extend SQLite support to 3.26+

Laravel currently requires SQLite 3.35+, this PR extends SQLite support to 3.26+

### Add and Drop Foreign Keys on SQLite
Fixes #25475, #23461. SQLite [doesn't support](https://www.sqlite.org/omitted.html#:~:text=Complete%20ALTER%20TABLE%20support) adding / dropping foreign keys with `alter table` command, but as suggested on [the docs](https://www.sqlite.org/lang_altertable.html#:~:text=7.%20Making%20Other%20Kinds%20Of%20Table%20Schema%20Changes), you may recreate the table with arbitrary changes, that's what this PR do to implement these commands.

This PR adds support for blueprint's `$table->foreign()`, `$table->dropForeign()` methods and `->constrained()` modifier on SQLite:

```php
Schema::table('posts', function (Blueprint $table) {
    $table->unsignedBigInteger('user_id');

    $table->foreign('user_id')->references('id')->on('users'); // Now works as expected on SQLite
});

Schema::table('posts', function (Blueprint $table) {
    $table->foreignId('user_id')
        ->constrained(); // Now works as expected on SQLite
});

Schema::table('posts', function (Blueprint $table) {
    $table->dropForeign(['user_id']); // Now works as expected on SQLite, no exception will be thrown anymore!
});
```

Fixes #51318, SQLite doesn't allow dropping a column [if it is used in a foreign key constraint](https://www.sqlite.org/lang_altertable.html#:~:text=The%20column%20is%20used%20in%20a%20foreign%20key%20constraint.), this PR makes it possible:

```php
Schema::table('sessions', function (Blueprint $table) {
    $table->dropForeign(['user_id']);
    $table->dropColumn('user_id');
});

// or with more magic...
Schema::table('sessions', function (Blueprint $table) {
    $table->dropConstrainedForeignId('user_id');
});
```

### Add and Drop the primary key on SQLite

This PR adds support for blueprint's `$table->primary()`, `$table->dropPrimary()` methods, `->primary()`, and `->primary(false)->change()` modifiers when **modifying table** on SQLite:

```php
Schema::table('posts', function (Blueprint $table) {
    $table->dropPrimary();              // Drop the current primary key
    $table->string('email');            // Add 'email' column
    $table->primary(['name', 'email']); // Add a composite primary key 
});

Schema::table('posts', function (Blueprint $table) {
    $table->uuid('id')->primary();  // Add 'id' column and make it primary
});

Schema::table('posts', function (Blueprint $table) {
    $table->string('id')->primary(false)->change();  // Change the type of 'id' column to `string` and drop primary key
});
```

### Preserve the order of commands

This is the most important change on this PR; When updating table and compiling blueprint commands to SQL, the order of commands wasn't respected:
* First compile all modified columns at one query
* Second compile all added columns at once
* Then compile other commands

As discussed on #50925, this will causes issue, unexpected results and leads to error most of the time when updating table. For example, the following code raises `The 'username' column already exists!` SQL error on all databases, because `add column` command was compiled before `rename column`:

```php
Schema::table('users', function (Blueprint $table) {
    $table->renameColumn('username', 'email');
    $table->string('username');
});
```

This PR fixes this by preserving the order of commands as declared on the blueprint. You don't have to call each command in a separate `Schema::table` anymore that breaks migration transaction!

### Blueprint State for SQLite

As explained before, when updating table on SQLite, we have to recreate the table to do arbitrary changes. To do so we need to know the current state of the table on each command. This PR adds `BlueprintState` class to keep track of changes on the table. After this PR we are able to mix any combination of blueprint commands in a single `Schema::table` (as demonstrated on [the added integration tests](https://github.com/laravel/framework/pull/51373/files#diff-1c59c08b17e52945b97edf1eba18a1381ae384e106749a159cc4144032c45413)) and it works as expected in the same order!

## Notes
1. Unlike other database drivers, foreign keys don't have name on SQLite, so obviously you can't drop a foreign key by its name on SQLite:

    ```php
    Schema::table('posts', function (Blueprint $table) {
        $table->dropForeign('posts_user_id_foreign'); // ❌ doesn't work on SQLite

        $table->dropForeign(['user_id']);             // ✔️ works as expected
    }
    ```
1. We probably need to remove [`withSqlite` trait on `orchestral/testbench` package](https://github.com/orchestral/testbench-core/blob/eabe0c72e787aae29df96bd4e33c61fffa6d5edc/src/Concerns/Database/WithSqlite.php), it seems we don't need that anymore if it's merged?

### CODE REVIEW

**Comment by shaedrich** [comment]
Is there a reason, you mixed the SQLite and migration execution order changes into one PR instead of two separate ones?

**Comment by ocleroux** [comment]
I'm eager to see this merged!

**Comment by hafezdivandari** [comment]
@shaedrich we can't add SQLite features without fixing the migration commands' orders.

**Comment by shaedrich** [comment]
@hafezdivandari You could have created the SQLite PR as draft, added a comment concerning the PR merging order and have backmerged the `main` branch after the migration command order PR was merged

**Comment by hafezdivandari** [comment]
@shaedrich It's fine ;)

**Comment by taylorotwell** [comment]
To be honest this just is too much for us to take on reviewing and maintaining at the moment and I'm not sure the return on investment is there for end users.

**Comment by driesvints** [comment]
Hi @hafezdivandari. I think we should still give a part of your PR a chance but like Taylor said and like we said on the Passport PR keep in mind to keep PR's small and separate because reviewing and testing such a large PR as this is simply too much for us at this time. Sending in incremental PR's with the different topics from this PR will be much much more easier to review.

I feel like the most important part of this bit is the dropping of foreign keys in SQLite. Could we maybe just focus on that one for now so we can get that back into Laravel v11?

Besides that I went through your code changes and while I feel most of this looks okay I do have a question why you decided to move column modifications to different `alter` lines instead of packing them into a single query statement with `add` like it does now. Because with this PR it could lead to a head more query statements. Not sure if that's particularly bad but it did caught my eye and I'd like to get your opinion about it.

**Comment by hafezdivandari** [comment]
Hi @driesvints,

> I think we should still give a part of your PR a chance but like Taylor said and like we said on the Passport PR keep in mind to keep PR's small and separate because reviewing and testing such a large PR as this is simply too much for us at this time. Sending in incremental PR's with the different topics from this PR will be much much more easier to review.

First of all, sorry for sending large PRs 😅 but this one is a little different from the PR on the Passport, separating this into more PRs (if technically possible) doesn't make it easier to review. Because although this PR fixes 2 main problems at the same time, it doesn't mean the changes can necessarily be sent on 2 separate PRs.

But I can add comments on each part of the code changes and explain in more details if you want. IMHO that would be really helpful when reviewing.

> I feel like the most important part of this bit is the dropping of foreign keys in SQLite. Could we maybe just focus on that one for now so we can get that back into Laravel v11?

As I tried to explain on [this comment earlier](https://github.com/laravel/framework/pull/51373#issuecomment-2107788660) and [this one](https://github.com/laravel/framework/pull/51358#issuecomment-2102857381), It is not possible to fix the drop foreign on SQLite (or any other commands that needs recreating the table on SQLite) without fixing the ["order of commands"](https://github.com/laravel/framework/pull/51373#:~:text=Preserve%20the%20order%20of%20commands) problem first.

> Besides that I went through your code changes and while I feel most of this looks okay I do have a question why you decided to move column modifications to different alter lines instead of packing them into a single query statement with add like it does now. Because with this PR it could lead to a head more query statements. Not sure if that's particularly bad but it did caught my eye and I'd like to get your opinion about it.

Because preserving the order of commands matters. Please read the "Preserve the order of commands" section above.

**Comment by driesvints** [comment]
We're gonna have another look at this but it could be a few weeks before we get to it. Thanks!

**Comment by gsxdsm** [comment]
FWIW I’m running into the issue of not being able to drop foreign keys in SQLite and would very much appreciate having this fix in. It’s making a mess of things on my end.

File: src/Illuminate/Database/Schema/Blueprint.php [hunk 1]
```diff
@@ -260,6 +277,48 @@ public function addFluentCommands(Connection $connection, Grammar $grammar)
...
+            $alterCommands = $grammar->getAlterCommands($connection);
+            $commands = [];
+            $lastCommandWasAlter = false;
+            $hasAlterCommand = false;
+
+            foreach ($this->commands as $command) {
+                if (in_array($command->name, $alterCommands)) {
+                    $lastCommandWasAlter = true;
+                    $hasAlterCommand = true;
+                } else {
```
Comment (imanghafoori1) [inline src/Illuminate/Database/Schema/Blueprint.php:299]: You can merge it into `elseif ($lastCommandWasAlter) {`

File: src/Illuminate/Database/Schema/Blueprint.php [hunk 2]
```diff
@@ -78,6 +79,13 @@ class Blueprint
      */
     public $after;
 
+    /**
+     * The blueprint state instance.
+     *
+     * @var \Illuminate\Database\Schema\BlueprintState
```
Comment (imanghafoori1) [inline src/Illuminate/Database/Schema/Blueprint.php:85]: you may add the `null` here:

@var \Illuminate\Database\Schema\BlueprintState|null

File: src/Illuminate/Database/Schema/Blueprint.php [hunk 3]
```diff
@@ -1740,6 +1807,27 @@ public function getCommands()
...
+
+    /*
+     * Determine if the blueprint has state.
+     *
+     * @param  mixed  $name
+     * @return bool
+     */
+    private function hasState(): bool
+    {
+        return ! is_null($this->state);
```
Comment (imanghafoori1) [inline src/Illuminate/Database/Schema/Blueprint.php:1828]: return  (bool) $this->state;

File: src/Illuminate/Database/Schema/Blueprint.php [hunk 4]
```diff
@@ -1671,6 +1734,10 @@ public function removeColumn($name)
             return $c['name'] != $name;
         }));
 
+        $this->commands = array_values(array_filter($this->commands, function ($c) use ($name) {
```
Comment (imanghafoori1) [inline src/Illuminate/Database/Schema/Blueprint.php:1737]: You may also use an arrow function here

File: src/Illuminate/Database/Schema/Blueprint.php (hunk 1 above)
Comment (hafezdivandari) [reply src/Illuminate/Database/Schema/Blueprint.php]: Thanks!

File: src/Illuminate/Database/Schema/Blueprint.php (hunk 3 above)
Comment (hafezdivandari) [reply src/Illuminate/Database/Schema/Blueprint.php:1826]: It's more clear as is?

File: src/Illuminate/Database/Schema/Blueprint.php (hunk 4 above)
Comment (hafezdivandari) [reply src/Illuminate/Database/Schema/Blueprint.php:1735]: Same style as the upper line, I think we should fix them both or neither.

File: src/Illuminate/Database/Schema/Blueprint.php (hunk 3 above)
Comment (imanghafoori1) [reply src/Illuminate/Database/Schema/Blueprint.php:1826]: `! is_null` is kinda double negation.

File: src/Illuminate/Database/Schema/Blueprint.php [hunk 5]
```diff
@@ -260,6 +277,46 @@ public function addFluentCommands(Connection $connection, Grammar $grammar)
...
+                    $commands[] = $this->createCommand('alter');
+                    $lastCommandWasAlter = false;
+                }
+
+                $commands[] = $command;
+            }
+
+            if ($lastCommandWasAlter) {
+                $commands[] = $this->createCommand('alter');
+                $lastCommandWasAlter = false;
```
Comment (imanghafoori1) [inline src/Illuminate/Database/Schema/Blueprint.php:309]: it seems that you can remove this line

File: src/Illuminate/Database/Schema/BlueprintState.php [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+     */
+    public function getColumns()
+    {
+        return $this->columns;
+    }
+
+    /**
+     * Get the primary key.
+     *
+     * @return null|\Illuminate\Database\Schema\IndexDefinition
```
Comment (imanghafoori1) [inline src/Illuminate/Database/Schema/BlueprintState.php:128]: \Illuminate\Database\Schema\IndexDefinition|null

File: src/Illuminate/Database/Schema/Blueprint.php (hunk 5 above)
Comment (imanghafoori1) [reply src/Illuminate/Database/Schema/Blueprint.php:309]: image
it seems that you can remove this line since the var is no longer used down the line.

File: src/Illuminate/Database/Schema/BlueprintState.php [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+     * @var \Illuminate\Database\Schema\Grammars\Grammar
+     */
+    protected $grammar;
+
+    /**
+     * The columns.
+     *
+     * @var \Illuminate\Database\Schema\ColumnDefinition[]
+     */
+    private $columns = [];
```
Comment (imanghafoori1) [inline src/Illuminate/Database/Schema/BlueprintState.php:38]: You may remove redundant initializers (`= [];`) like below since they are assigned immediately in the constructor. 

`private $columns;`

File: src/Illuminate/Database/Schema/BlueprintState.php [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+        $this->columns = collect($schema->getColumns($table))->map(fn ($column) => new ColumnDefinition([
+            'name' => $column['name'],
+            'type' => $column['type_name'],
+            'full_type_definition' => $column['type'],
+            'nullable' => $column['nullable'],
+            'default' => is_null($column['default']) ? null : new Expression($column['default']),
+            'autoIncrement' => $column['auto_increment'],
+            'collation' => $column['collation'],
+            'comment' => $column['comment'],
+            'virtualAs' => ! is_null($column['generation']) && $column['generation']['type'] === 'virtual'
```
Comment (imanghafoori1) [inline src/Illuminate/Database/Schema/BlueprintState.php:87]: Shorter form:
`
  'virtualAs' =>($column['generation']['type'] ?? '') === 'virtual'
                ? $column['generation']['expression'] : null,
 'storedAs' => ($column['generation']['type'] ?? '') === 'stored'
                ? $column['generation']['expression'] : null,
`

Comment (hafezdivandari) [reply src/Illuminate/Database/Schema/BlueprintState.php:87]: It's just moved from here: https://github.com/laravel/framework/pull/51373/files#diff-7cf2f940b724463282b1dbae24464cb3a7c7c0864d40b4c3706eb99f65994d2cL279

File: src/Illuminate/Database/Schema/BlueprintState.php (hunk 7 above)
Comment (hafezdivandari) [reply src/Illuminate/Database/Schema/BlueprintState.php]: Thanks!

File: src/Illuminate/Database/Schema/BlueprintState.php [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+     * The columns.
+     *
+     * @var \Illuminate\Database\Schema\ColumnDefinition[]
+     */
+    private $columns = [];
+
+    /**
+     * The primary key.
+     *
+     * @var null|\Illuminate\Database\Schema\IndexDefinition
```
[the remark this answers is not in the document]
Comment (hafezdivandari) [reply src/Illuminate/Database/Schema/BlueprintState.php]: Thanks!

File: src/Illuminate/Database/Schema/BlueprintState.php (hunk 6 above)
Comment (hafezdivandari) [reply src/Illuminate/Database/Schema/BlueprintState.php]: Thanks!

File: src/Illuminate/Database/Schema/Blueprint.php (hunk 5 above)
Comment (hafezdivandari) [reply src/Illuminate/Database/Schema/Blueprint.php:309]: IDK I prefer to keep it as is.

Comment (imanghafoori1) [reply src/Illuminate/Database/Schema/Blueprint.php:309]: I think someone will come up with PR to remove the unused variable at some point. No reason to define it.
I mean line 309 `$lastCommandWasAlter = false;`

**Comment by taylorotwell** [comment]
So, I've reviewed this PR. I do have a few concerns. First, given my last 14 years of experience managing this framework, I feel like it's a 99.9% chance this PR will break existing applications given its size and nature of changes. Do we have anyone with large codebases that can try it?

Secondly, I fear for my ability to maintain it going forward. @hafezdivandari can you summarize the changes and how they work in plain English?

File: src/Illuminate/Database/Schema/Blueprint.php [hunk 10]
```diff
@@ -180,17 +194,20 @@ protected function commandsNamed(array $names)
...
     protected function addImpliedCommands(Connection $connection, Grammar $grammar)
     {
-        if (count($this->getAddedColumns()) > 0 && ! $this->creating()) {
-            array_unshift($this->commands, $this->createCommand('add'));
-        }
-
-        if (count($this->getChangedColumns()) > 0 && ! $this->creating()) {
-            array_unshift($this->commands, $this->createCommand('change'));
-        }
-
```
Comment (hafezdivandari) [inline src/Illuminate/Database/Schema/Blueprint.php:190]: Step 2. We have already added `ColumnDefinition`s to list of Blueprint's commands, so there is no need to prepend a command for added / changed columns.

File: src/Illuminate/Database/Schema/Blueprint.php [hunk 11]
```diff
@@ -180,17 +194,20 @@ protected function commandsNamed(array $names)
...
+        if (! $this->creating()) {
+            $this->commands = array_map(
+                fn ($command) => $command instanceof ColumnDefinition
+                    ? $this->createCommand($command->change ? 'change' : 'add', ['column' => $command])
+                    : $command,
+                $this->commands
+            );
+
+            $this->addAlterCommands($connection, $grammar);
+        }
```
Comment (hafezdivandari) [inline src/Illuminate/Database/Schema/Blueprint.php:210]: Step 3. We walk through the commands and convert `ColumnDefinition`s into eighter an "add" or a "change" command.

Then we call `addAlterCommands` method.

File: src/Illuminate/Database/Schema/Blueprint.php [hunk 12]
```diff
@@ -260,6 +277,48 @@ public function addFluentCommands(Connection $connection, Grammar $grammar)
...
     }
 
+    /**
+     * Add the alter commands if whenever needed.
+     *
+     * @param  \Illuminate\Database\Connection  $connection
+     * @param  \Illuminate\Database\Schema\Grammars\Grammar  $grammar
+     * @return void
+     */
+    public function addAlterCommands(Connection $connection, Grammar $grammar)
```
Comment (hafezdivandari) [inline src/Illuminate/Database/Schema/Blueprint.php:287]: Step 4. SQLite doesn't support some alter commands, so we need to recreate the table in order to apply changes.
This method walks through the commands and adds "alter commands" whenever needed.
If any alter command has been added, we initiate an "state" for the Blueprint.

File: src/Illuminate/Database/Schema/Blueprint.php [hunk 13]
```diff
@@ -136,6 +144,10 @@ public function toSql(Connection $connection, Grammar $grammar)
             $method = 'compile'.ucfirst($command->name);
 
             if (method_exists($grammar, $method) || $grammar::hasMacro($method)) {
+                if ($this->hasState()) {
+                    $this->state->update($command);
+                }
+
```
Comment (hafezdivandari) [inline src/Illuminate/Database/Schema/Blueprint.php:150]: Step 5. When compiling to SQL, we update the state on each command.

**Review by hafezdivandari — COMMENTED** [review]
### Summarized List of changes 
1. All the changes are about **modifying** tables and nothing has changed when creating tables.
2. When adding or modifying columns, we were compiling all the changed columns at once, then compiling all the added columns and then compiling other commands. But we are now compiling each command one by one and in order.
3. Some commands were not supported by SQLite, and we need to recreate the table in order to apply these commands. The "alter commands" will be added to list of `Blueprint` commands whenever needed. The new `BlueprintState` class helps us keep the current state of the table to be able to recreate it.

### Breaking changes?
 1. Although this PR changes the order of commands, but the result should be the same, because it wasn't possible to call dependent commands in a single Blueprint before this PR. Any counterexample?
 2. Some packages (e.g DB drivers) have overridden the `compileAdd` / `compileChange` methods of the `Schema\Grammars\Grammar` class. The same changes should have been applied to these overridden methods as well to avoid breaking changes. Let me know if you need help on this.

### Ability to maintain?
1. TBH, one reason for sending one large PR instead of several small ones, was to be able to revert the whole thing easily if needed!
2. It's now much easier to understand how schema builder works - each command compiles to a single query in the same expected order!
3. If SQLite adds support for any of the alter commands in its next version, this would be very easy for us to add support for that on the framework too.

**Comment by taylorotwell** [comment]
Thanks

**Comment by crynobone** [comment]
Hi @hafezdivandari 

1. We probably need to remove [withSqlite trait on orchestral/testbench package](https://github.com/orchestral/testbench-core/blob/eabe0c72e787aae29df96bd4e33c61fffa6d5edc/src/Concerns/Database/WithSqlite.php), it seems we don't need that anymore if it's merged?

`WithSqlite` is not being used anywhere by default, and requires explicit invoke to trigger.

2. I started seeing some error with using `Schema:::dropIfExists()` with in-memory SQLite after the latest upgrade on Testbench.

```
1) Orchestra\Testbench\Tests\Databases\RefreshDatabaseUsingEventsTest::it_create_database_migrations
Error: Call to a member function getAttribute() on null

/Projects/orchestra/testbench-core/vendor/laravel/framework/src/Illuminate/Database/Connection.php:1632
/Projects/orchestra/testbench-core/vendor/laravel/framework/src/Illuminate/Database/Schema/Grammars/SQLiteGrammar.php:39
/Projects/orchestra/testbench-core/vendor/laravel/framework/src/Illuminate/Database/Schema/Blueprint.php:293
/Projects/orchestra/testbench-core/vendor/laravel/framework/src/Illuminate/Database/Schema/Blueprint.php:209
/Projects/orchestra/testbench-core/vendor/laravel/framework/src/Illuminate/Database/Schema/Blueprint.php:130
/Projects/orchestra/testbench-core/vendor/laravel/framework/src/Illuminate/Database/Schema/Blueprint.php:116
/Projects/orchestra/testbench-core/vendor/laravel/framework/src/Illuminate/Database/Schema/Builder.php:564
/Projects/orchestra/testbench-core/vendor/laravel/framework/src/Illuminate/Database/Schema/Builder.php:446
/Projects/orchestra/testbench-core/vendor/laravel/framework/src/Illuminate/Support/Facades/Facade.php:357
/Projects/orchestra/testbench-core/tests/Databases/RefreshDatabaseUsingEventsTest.php:36
/Projects/orchestra/testbench-core/src/Concerns/HandlesDatabases.php:47
/Projects/orchestra/testbench-core/src/Concerns/ApplicationTestingHooks.php:267
/Projects/orchestra/testbench-core/src/Concerns/ApplicationTestingHooks.php:110
/Projects/orchestra/testbench-core/src/Concerns/Testing.php:69
/Projects/orchestra/testbench-core/vendor/laravel/framework/src/Illuminate/Collections/helpers.php:236
/Projects/orchestra/testbench-core/src/functions.php:119
/Projects/orchestra/testbench-core/vendor/laravel/framework/src/Illuminate/Collections/helpers.php:236
/Projects/orchestra/testbench-core/src/Concerns/Testing.php:97
/Projects/orchestra/testbench-core/src/TestCase.php:65
```

https://github.com/orchestral/testbench-core/actions/runs/9866136206

**Comment by hafezdivandari** [comment]
@crynobone the error is on [this line](https://github.com/laravel/framework/blob/68e88bbcee1e9b0b02ca267916ee25f45ab3935c/src/Illuminate/Database/Connection.php#L1632), it seems that connection's `getPdo()` is returning `null` which is wierd. I'm still investigating.

**Comment by hafezdivandari** [comment]
@crynobone I may have found the cause of issue:

The `RefreshDatabase` trait [calls `$connection->disconnect()`](https://github.com/laravel/framework/blob/68e88bbcee1e9b0b02ca267916ee25f45ab3935c/src/Illuminate/Foundation/Testing/RefreshDatabase.php#L111) so the connection's pdo is null on `destroyDatabaseMigrations()` method.

It seems that schema builder doesn't reconnect the connection!

**Comment by crynobone** [comment]
Noted. Probably something Testbench specific. I will investigate further. Thanks @hafezdivandari

**Comment by michaeldyrynda** [comment]
Adding a note here that this merge change appears to have affected some tests in my [laravel-model-uuid](https://github.com/michaeldyrynda/laravel-model-uuid) package - specifically where we're [making assertions](https://github.com/michaeldyrynda/laravel-model-uuid/actions/runs/9865955429/job/27243869091) against the shape of the resulting alter statements.

As I'm specifically testing that the generated queries are correctly formatted, I don't necessarily think this is a breaking change to end users as the resulting query is ultimately the same, just that now there are multiple queries rather than a single one.

Not sure this has tangible impact to framework consumers, though.

**Comment by hafezdivandari** [comment]
Thanks @michaeldyrynda, these changes were expected, as I explained on [this comment](https://github.com/laravel/framework/pull/51373#pullrequestreview-2157957688) before merge and [the changes on framework's tests](https://github.com/laravel/framework/pull/51373/files#diff-ba5578efde828aa6d2dfeb7907fac68ae6bb840bbf81f0df4931f367eae61916) also show that.

Nice package by the way. You may also want to check PR #50355 for use in [your code](https://github.com/michaeldyrynda/laravel-model-uuid/blob/05fd05d5bf680d32d30aa2a1fed3f9b2cfe3858e/src/LaravelModelUuidServiceProvider.php#L28-L40).

**Comment by hafezdivandari** [comment]
@hafezdivandari off course Taylor's feeling is right. It was a hard decision to make and the PR was open for 2 months.

Besides that, it was much easier for me to send this to 12.x. Since the patch release on Tuesday, I'm checking the issue board every hour and try to find related discussions on other packages too!

> However, finding the root cause took quite some time, because this change wasn't appropriately documented (due to the size of the PR?) and it's breaking character wasn't obvious at first.

The PR description couldn't be more long I think, tried to cover every aspect + other comments on this thread. But I believe this maybe could be covered better on weekly Laravel newsletters after release to better inform the community.

**Comment by fulopattila122** [comment]
I think this might be a breaking change. Disclaimer: the feature set is nice, so I'm not complaining, just thought to share the information.

One of our migrations started to fail on SQLite, beginning with Laravel 11.15:
https://github.com/vanilophp/framework/blob/4.1.0/src/Links/resources/database/migrations/2024_05_31_074502_add_root_item_to_link_group_items_table.php

Having Laravel 11.15 it fails:

image

Switching back to Laravel 11.14 solves the problem:

image

The linked example is from a library that supports Laravel 10.43+ and Laravel 11.0+

I believe we can find a workaround, like checking what Laravel version we have, and acting accordingly. But maybe you have a better suggestion.

**Comment by ekandreas** [comment]
Ohh, this is a large one... Breaking the tests for us.
Are there any more simple workarounds?

**Comment by hafezdivandari** [comment]
@ekandreas would you please share more info? Code? Exception?

**Comment by ekandreas** [comment]
Sorry, some kind of integrity violation in our seeder that caused an error.
We had to exclude some foreign keys in the schemas.
```
            if (env('DB_CONNECTION') == 'sqlite') {
                return;
            }
```

File: src/Illuminate/Database/Schema/Grammars/SQLiteGrammar.php [hunk 14]
```diff
@@ -28,6 +26,23 @@ class SQLiteGrammar extends Grammar
...
+     * Get the commands to be compiled on the alter command.
+     *
+     * @param  \Illuminate\Database\Connection  $connection
+     * @return array
+     */
+    public function getAlterCommands(Connection $connection)
+    {
+        $alterCommands = ['change', 'primary', 'dropPrimary', 'foreign', 'dropForeign'];
+
+        if (version_compare($connection->getServerVersion(), '3.35', '<')) {
```
Comment (mjauvin) [inline src/Illuminate/Database/Schema/Grammars/SQLiteGrammar.php:39]: @hafezdivandari shouldn't the operator be ">=", here? If the goal is to make sure SQLite >= 3.35 uses the `dropColumn` command instead of recreating the table?

Comment (hafezdivandari) [reply src/Illuminate/Database/Schema/Grammars/SQLiteGrammar.php:39]: No, this method prepares an array of commands that are not supported by SQLite, and we have to recreate the table as a workaround for them. So as dropping column is not supported on <3.35 we add `dropColumn` command to the list.
